proposals/2024-06-05_Track_Similarity_Column.md

```diff
@@ -0,0 +1,168 @@
+algorithm for choosing the Sync Leader.
+
+We can provide context menu options to target other decks, or tracks from the library.
+
+### Computing Compatibility
+
+Python Pseudocode:
+
+```python
+import math
+
+# Track1: Owl City - Lucid Dream
+key1 = 3 # Eb, if C is 0, C# is 1 and so on...
+bpm1 = 128
+# Track2: Alan Walker - Alone
+key2 = 10 # Bb
+bpm2 = 97
+
+def pitch_delta_on_tempo_stretching(old_bpm, new_bpm):
+  return math.log2(new_bpm / old_bpm) * 12 # change of pitch in semitones
+
+def normalize_pitch_value(val):
+  return (val + 12) % 12 # confine the pitch to the [0-12) range
+
+def pitch_diff(key1, bpm1, key2, bpm2):
+  # calculate the pitch delta for each track when stretched to a tempo of 100BPM
+  delta1 = pitch_delta_on_tempo_stretching(bpm1, 100)
+  delta2 = pitch_delta_on_tempo_stretching(bpm2, 100)
+
+  # get the resulting key for each track at 100BPM
+  res_key1 = normalize_pitch_value(key1 + delta1)
+  res_key2 = normalize_pitch_value(key2 + delta2)
+
+  # now return the pitch difference when both tracks are played at the same tempo
+  return normalize_pitch_value(res_key1 - res_key2)
+
+pd = pitch_diff(key1, bpm1, key2, bpm2)
+print("Pitch Difference:", pd) # Output: 0.19895410624553378
+
+```
+
+For the examples above, we find that both these tracks are almost in the same key (with a 20 cent difference)
+when they’re played at the same tempo.We would not have known this by looking at the keys or the BPMs independently.
+This method of calculating harmonic compatibility gives users a deeper insight on which tracks will work well.
+
+Harmonically compatible tracks do not need to end up in the same key. We can account for this using the Circle of Fifths.
+Here’s the continued Python pseudocode:
+
+```python
+circle_of_fifths = [
+    0,  # C
+    7,  # G
+    2,  # D
+    9,  # A
+    4,  # E
+    11,  # B
+    6,  # Gb
+    1,  # Db
+    8,  # Ab
+    3,  # Eb
+    10,  # Bb
+    5,  # F
+]
+
+def circle_of_fifths_distance(note):
+    # distance of given note from note 0
+    # it's a circle, so the shortest distance may be from the end of the array
+    return min(circle_of_fifths.index(note), 12 - circle_of_fifths.index(note))
+
+def compatibility(pitch_difference):
+    rounded_pitch_diff = round(pitch_difference)
+    delta = abs(rounded_pitch_diff - pitch_difference)
+    # get a rank from 0 to 6, lower means more compatible
+    cof_rank = circle_of_fifths_distance(rounded_pitch_diff)
+    return cof_rank, delta
+
+comp = compatibility(pd)
+print("Rank:", comp[0], "Delta:", comp[1]) # Rank: 0 Delta: 0.19895410624553378
+```
+
+This gives us a rank, where 0 means the tracks will be in the same key,
+and 6 means the tracks will sound most dissonant together.
+
+We also get a delta/detune value to find out how off tune the tracks will be in cents.
+
+TODO: The algorithm needs to be able to take in tuning information, for non 440Hz tuning standards.
```

Comment:
I think this is something we can consider out of scope for the first version.  As always, we can take it into account so that a future version could possibly handle it, but I think trying to support this from the start will add a lot of unnecessary complexity.

Reply:
All the tracks are analyzed with 440 Hz. A 432 Hz Track is "just" a slightly pitched track compared to others. Since the algorithm needs to deal with pitched tracks anyway, this part comes for free. We just must not prevent it artificially.  

The missing bit is to define the library storage format and add a GUI to enter the tuning info. This is out of scope of this project. 

We can already read Tracks annotated with keys like "A#m +50" but we discard the tuning info currently. Once this is "fixed" we can support tuning: https://github.com/mixxxdj/mixxx/blob/3d8e7fec8ba03fac34769bb2de6e863fd67b3efa/src/track/keyutils.cpp#L329

Reply:
sounds good to me